lib/Conversion/TritonGPUToLLVM/TargetInfoBase.cpp

```diff
@@ -0,0 +1,81 @@
+#include "triton/Conversion/TritonGPUToLLVM/TargetInfoBase.h"
+#include "triton/Conversion/TritonGPUToLLVM/Utility.h"
+
+namespace mlir::triton {
+
+std::optional<int>
+TargetInfoBase::getWarpGroupStartThreadId(Block *block) const {
+  using namespace triton::gpu;
+
+  // Look for an enclosing `ttg.warp_specialize` op.
+  while (block && block->getParentOp() &&
+         !isa<WarpSpecializePartitionsOp>(block->getParentOp()))
+    block = block->getParentOp()->getBlock();
+  if (!block || !block->getParentOp())
+    return {};
+
+  auto partitions = cast<WarpSpecializePartitionsOp>(block->getParentOp());
+  unsigned idx = block->getParent()->getRegionNumber();
+  WarpSpecializeOp ws = partitions.getParentOp();
+  std::optional<ArrayRef<int32_t>> startIds = ws.getWarpGroupStartIds();
+  assert(startIds && "cannot get warp group ID before warp group allocation");
+  int32_t warpStartId = (*startIds)[idx];
+  int threadsPerWarp =
+      TritonGPUDialect::getThreadsPerWarp(ws->getParentOfType<ModuleOp>());
+  return warpStartId * threadsPerWarp;
+}
+
+Value TargetInfoBase::getThreadId(RewriterBase &rewriter, Location loc) const {
```

Comment:
so both backend call into this one? What is the point of moving it to TargetInfo then?

Reply:
It's not about thread ID actually--we can share that. It's about wave ID as mentioned in the pull request message. We (will) customize that based on different architectures. On gfx12 for example, we can directly use `llvm.amdgpu.wave.id`; don't need to deduce from thread ID. Some more backgrouds in https://github.com/llvm/llvm-project/issues/163835.

I guess another way is to introduce a `ttg.warp_id` or something so that we can emit that and each backend can have different lowering for that. Though the question is whether it should exist in ttg dialect or some other dialect. Or maybe upstream GPU dialect..

Reply:
I think that's fine but we really should move to a different dialect at some point

I can handle it later though

Reply:
Let's parallel then--I'll ask somebody to create such a dialect and move the existing `ttg.local_barrier` into it. Do you have a particular naming consideration for this dialect? ttg is already taken. Maybe ttw as shorthand of "triton warp" level concepts?